File: src/hey/LongestPalindromicSubstring.java
/*
    problem: Median of Two Sorted Arrays
    author: everpuck
    date: 02/10/2019
*/

class Solution {
    public String longestPalindrome(String s) {
        int length = s.length();
        int maxIndex = 0;
        int maxLength = 0;
        boolean isOdd = true;
        for (int i = 0; i < length; ++i) {
            for (int j = 0; j < i; ++j) {
                if (i + j + 1 >= length || s.charAt(i + j + 1) != s.charAt(i - j - 1)) {
                    break;
                }
                if (j + 1 > maxLength) {
                    maxLength = j + 1;
                    maxIndex = i;
                    isOdd = true;
                }
            }
            if (i + 1 < length && s.charAt(i) == s.charAt(i+1)) {
                for (int j = 0; j < i; ++j) {
                    if (i + j + 2 >= length || s.charAt(i + j + 2) != s.charAt(i - j - 1)) {
                        break;
                    }
                    if (j + 1 >= maxLength) {
                        maxLength = j + 1;
                        maxIndex = i;
                        isOdd = false;
                    }
                }
            }
            if (i + 1 < length && s.charAt(i) == s.charAt(i+1) && maxLength < 1) {
                maxIndex = i;
                isOdd = false;
            }
        }
        if (length < 1) {
            return "";
        }
        if (isOdd) {
            return s.substring(maxIndex - maxLength, maxIndex + maxLength + 1);
        }
        else {
            return s.substring(maxIndex - maxLength, maxIndex + maxLength + 2);
        }
        
    }
}


public class LongestPalindromicSubstring {
    public static void main(String[] args) {
        String retString = "";
        // String s = new String("babad");
        // String s = new String("cbbd");
        // String s = new String("aaaabaa");
        String s = new String("");
        Solution solution = new Solution();
        retString = solution.longestPalindrome(s);
        System.out.println(retString);
    }
}

File: src/hey/LongestPalindromicSubstring.py
# problem: Longest Palindromic Substring 
# author: everpuck
# date: 2/2/2019


class Solution:
    def longestPalindrome(self, s):
        """
        :type s: str
        :rtype: str
        """
        if not s or len(s) < 2:
            return s
        
        max_index = 0
        max_length = 0
        is_odd = True
        length = len(s)
        for i in range(length):

            for j in range(i):
                if i + j + 1>= length or s [i + j + 1] != s[i - j -1]:
                    break
                if j + 1 > max_length:
                    max_length, max_index , is_odd = j + 1, i, True
            
            if i + 1 < length and s[i] == s[i+1]:
                for j in range(i):
                    if i + j + 2 >= length or s [i + j + 2] != s[i - j - 1]:
                        break
                    if j + 1 >= max_length:
                        max_length, max_index, is_odd = j + 1, i, False
            # elif i + 1 < length and s[i] != s[i + 1]:
            #     continue
        
            if i + 1 < length and s[i] == s[i+1] and max_length < 1:
                max_index, is_odd = i, False

        if is_odd:
            ret = s[max_index - max_length: max_index + max_length + 1]
        else:
            ret = s[max_index - max_length: max_index + max_length + 2]
        
        return ret


if __name__ == '__main__':
    s = 'sbabbad'
    # s = 'bbd'
    s = 'aaaa'
    # s = "aaabaaaa"
    # s = 'aaaabaa'
    solution = Solution()
    print(solution.longestPalindrome(s))
